### PR TITLE
[docs] Add `DateRangePicker` example with a `Button` trigger

### PR DESCRIPTION
**Pull request by LukasTy**
Improve based on https://github.com/mui/mui-x/issues/10469

### CODE REVIEW

**Comment by LukasTy** [comment]
> For Joy and the browser input we did a single demo for all pickers Do you think that for the button it's worth splitting it in two?

The existing Joy UI and browser input demos are basically unusable.
I tried navigating them a few times and it's not easy for me, I can't imagine what our end-users feel when checking those demos in detail. 🤔 
Of course, you can always just forget the expanded code option and go straight to Codesandbox, then the code is bit more digestible...

WDYT? 🤔 
I'm pretty sure I know what @noraleonte would say in this regard. 🙈 😆

**Comment by flaviendelangle** [comment]
> Of course, you can always just forget the expanded code option and go straight to Codesandbox, then the code is bit more digestible...

For this kind of big demos, I was indeed focusing on the codesandbo (or copy paste in IDE).
I'm not against the idea of splitting, but I guess that if we do, we should unify the existing demos in a follow up

**Comment by noraleonte** [comment]
Just catching up with this 👁️ 

I like the idea of splitting this, I already feel like the previous demos are too heavy, so maybe they are not a very good reference. I also agree that if we split here (which IMO we should) we should split the other ones in a follow-up.

I am very reluctant to rely JUST on codesandbox or copy&paste... If we rely on the fact that people will open codesandbox or an IDE, we exclude users who want to just quickly check an implementation example. I doubt **all** users will bother to go through the extra step(s). Also, hard to digest code is hard to digest code regardless of the environment, IMHO.

File: docs/data/date-pickers/custom-field/DateRangePickerWithButtonField.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+export default function DateRangePickerWithButtonField() {
+  const [value, setValue] = React.useState([null, null]);
+
+  return (
+    <LocalizationProvider dateAdapter={AdapterDayjs}>
+      <ButtonDateRangePicker
+        label={
+          value[0] === null && value[1] === null
+            ? null
+            : value.map((v) => (v ? v.format('MM/DD/YYYY') : 'null')).join(' - ')
```
Comment (noraleonte) [inline docs/data/date-pickers/custom-field/DateRangePickerWithButtonField.js:63]: Nitpick: I also have the tendency to use short variable names in maps, but I'm wondering if we should do that in demos 🤔 🙈 

```suggestion
            : value.map((date) => (date ? date.format('MM/DD/YYYY') : 'null')).join(' - ')
```

**Review by noraleonte — APPROVED** [review]
Great improvement! 🎉

File: docs/data/date-pickers/custom-field/DateRangePickerWithButtonField.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+export default function DateRangePickerWithButtonField() {
+  const [value, setValue] = React.useState<DateRange<Dayjs>>([null, null]);
+
+  return (
+    <LocalizationProvider dateAdapter={AdapterDayjs}>
+      <ButtonDateRangePicker
+        label={
+          value[0] === null && value[1] === null
+            ? null
+            : value.map((v) => (v ? v.format('MM/DD/YYYY') : 'null')).join(' - ')
```
Comment (LukasTy) [inline docs/data/date-pickers/custom-field/DateRangePickerWithButtonField.tsx:82]:
```suggestion
            : value.map((date) => (date ? date.format('MM/DD/YYYY') : 'null')).join(' - ')
```

File: docs/data/date-pickers/custom-field/DateRangePickerWithButtonField.tsx.preview [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<ButtonDateRangePicker
+  label={
+    value[0] === null && value[1] === null
+      ? null
+      : value.map((v) => (v ? v.format('MM/DD/YYYY') : 'null')).join(' - ')
```
Comment (LukasTy) [inline docs/data/date-pickers/custom-field/DateRangePickerWithButtonField.tsx.preview:5]:
```suggestion
      : value.map((date) => (date ? date.format('MM/DD/YYYY') : 'null')).join(' - ')
```

File: docs/data/date-pickers/custom-field/DateRangePickerWithButtonField.js (hunk 1 above)
Comment (LukasTy) [reply docs/data/date-pickers/custom-field/DateRangePickerWithButtonField.js:63]: Thanks for noticing it. 👍

**Comment by LukasTy** [comment]
> I'm not against the idea of splitting, but I guess that if we do, we should unify the existing demos in a follow up

> I like the idea of splitting this, I already feel like the previous demos are too heavy, so maybe they are not a very good reference. I also agree that if we split here (which IMO we should) we should split the other ones in a follow-up.

Splitting those demos would be a bit different. In this case, the root `Field` component has some differences, hence, it doesn't make much sense in having duplicate copies of the same root component.
But in case of those demos, the root `BrowserField` can be the same for all 3 Pickers and thus we would be duplicating ~50 lines of code in each of the demos creating the same `BrowserField`. 🤔 

Do these facts impact your opinion in any way @flaviendelangle @noraleonte?

**Comment by noraleonte** [comment]
@LukasTy those are very good points! 
I think that splitting those examples would inevitably cause some sort of duplication - they probably ended up in the same file to avoid that in the first place. But the more we work on them, the heavier they become, and at some point I would argue that an easy-to-understand file with duplicated root component is better than  a 400+ lines example file 🤔 

In case the joy and browser examples cannot be split in a way that the resulting files are more digestible, we should at least have a callout to signal that opening those files in a codesandbox would provide a much better experience 🤔 

Either way, I am fine with keeping the Joy and Browser examples as they are for the time being, but the complexity we currently have there should definitely be kept in mind in the future 💯 If we manage to split them, amazing 🎉 , if not, they probably should not be taken as a reference for the future demos because they fit a unique scenario. Making the future examples just as hard to read only for consistency doesn't make too much sense IMHO

**Comment by LukasTy** [comment]
@noraleonte Would you still agree that we should merge this based on the result of https://github.com/mui/mui-x/pull/10559 which can become a follow-up? 🤔

**Pull request merged**